src/wasm/wasm-validator.cpp

```diff
@@ -3970,6 +3970,33 @@ static void validateBinaryenIR(Module& wasm, ValidationInfo& info) {
 
 // Main validator class
 
+static void validateTypes(Module& module, ValidationInfo& info) {
+  // Check that public types do not contain any exact references if custom
+  // descriptors is not enabled. If they did, we would erase the exactness
+  // during binary writing and change the public type identities.
+  if (module.features.hasCustomDescriptors()) {
+    return;
+  }
+  for (auto type : ModuleUtils::getPublicHeapTypes(module)) {
```

Comment:
It looks like this adds a scan of the entire module in the common case (no CD, and validation is enabled)?

Perhaps we can avoid that overhead by instead erroring in the binary writer? That code already scans the module for heap types, and we could reuse that scan.

Reply:
I think the validator is a more principled location for this check. I'll investigate reimplementing `getPublicHeapTypes` to scan only the imports and exports and the types they can reach, though.

Reply:
I assume with that optimization this is practically a no-op now?

Reply:
Well it still has to scan the imports and exports and do an arbitrarily large traversal of the type graph, but at least it's not scanning the whole module now, so I wouldn't expect this to be a significant performance problem.